packages/devtools_app/lib/src/screens/vm_developer/object_viewport.dart

```diff
@@ -0,0 +1,118 @@
+
+/// Displays the VM information for the currently selected object in the
+/// program explorer.
+class ObjectViewport extends StatelessWidget {
+  ObjectViewport({
+    Key? key,
+    required this.controller,
+    required this.objectHistory,
+    this.initialObject,
+  }) : super(key: key);
+
+  static double get rowHeight => scaleByFontFactor(20.0);
+  static double get assumedCharacterWidth => scaleByFontFactor(16.0);
+
+  final ProgramExplorerController controller;
+  final ObjRef? initialObject;
+  final ObjectHistory objectHistory;
+  ValueListenable<bool> get refreshing => _refreshing;
+  final _refreshing = ValueNotifier<bool>(true);
+
+  @override
+  Widget build(BuildContext context) {
+    return ValueListenableBuilder(
+      valueListenable: _refreshing,
+      builder: (context, refreshing, _) {
+        return HistoryViewport<ObjRef>(
+          history: objectHistory,
+          controls: [
+            ToolbarAction(
+              icon: Icons.refresh,
+              onPressed: () {
+                _refreshing.value = !_refreshing.value;
+              },
+              tooltip: 'Refresh',
+            )
+          ],
+          generateTitle: (currentObjRef) {
+            return currentObjRef == null
+                ? 'No object selected.'
+                : _getViewportTitle(currentObjRef);
+          },
+          contentBuilder: (context, _) {
+            final currentObjRef = objectHistory.current.value;
+            return currentObjRef == null
+                ? const SizedBox.shrink()
+                : _buildObjectScreen(currentObjRef);
+          },
+        );
+      },
+    );
+  }
+}
+
+String _getViewportTitle(ObjRef objRef) {
+  if (objRef is ClassRef) return objRef.type + ' ' + (objRef.name ?? '<name>');
+  if (objRef is FuncRef) return objRef.type + ' ' + (objRef.name ?? '<name>');
+  if (objRef is FieldRef) return objRef.type + ' ' + (objRef.name ?? '<name>');
+  if (objRef is LibraryRef)
+    return objRef.type + ' ' + (objRef.name ?? '<name>');
+  if (objRef is ScriptRef) return 'Script @ ' + (objRef.uri ?? '<name>');
+  return '<unrecognized object>';
+}
+
+//Calls the object VM statistics card builder according to the VM Object type.
+Widget _buildObjectScreen(ObjRef objRef) {
+  if (objRef is ClassRef) {
+    return VmClassScreen(
+      controller: ClassScreenController(objRef),
+    );
+  }
+  if (objRef is FuncRef)
+    return const VMInfoCard(title: 'TO-DO: Display Function object data');
+  if (objRef is FieldRef)
+    return const VMInfoCard(title: 'TO-DO: Display Field object data');
+  if (objRef is LibraryRef)
+    return const VMInfoCard(title: 'TO-DO: Display Library object data');
+  if (objRef is ScriptRef)
+    return const VMInfoCard(title: 'TO-DO: Display Script object data');
+  return const SizedBox.shrink();
+}
+
+/// Manages the history of selected ObjRefs to make them accessible on a HistoryViewport.
+class ObjectHistory extends HistoryManager<ObjRef> {
```

Comment:
I don't see where `hasObjects` and `openedObjects` are used in this PR. Can we remove them until we have a function for these two getters?